### ISSUE TITLE
Bug Report: Parser sometimes fails to parse correct code at the end of a block/file

### ISSUE THREAD

**Issue by harrand** [harrand/psyc#2]
**Describe the bug**
Sometimes, the parser may fail to parse correct atoms. I think this is to do with the parser wanting more subtokens than it needs, and giving up early coz its ran out of subtokens even though the code is correct.

**Reproduction**
Build the following code:
```
main :: () -> i64
{
	puts("hello world! ;)");	
	greeting : string;
	defer greeting.cleanup();

	greeting.resize(5);
	puts(greeting.str);
	greeting.set("hey!");
	puts(greeting.str);

	print();
	return 0;
}

print :: () -> u0
{
	puts("free-function print!");
}

== default : build ==
{
	set_linkage_type(executable);
	set_output_name("scratchpad");
	use_stdlib(true);
}

//
```
remove the extra slashes/whitespace at the end and see repro.

**Expected behavior**
The code should cleanly compile (remove uses of string to break dependency on stdlib)

**Conditions (please complete the following information):**
 - OS: windows
 - Build Config: debug
 - Version: master (585e571e3624c251d2eb8b277b304ce65c793e59)
 - Language: psy
 - Compiler: psyc

**Comment by harrand** [comment]
At: https://github.com/harrand/psyc/commit/c10a1c25f23d377ec11c03973c64be8302219875#diff-ffaf3e3f67c72dd2f141e5bb3d0dfb861315cebe067eeec981d888b506a021daR1647

adding the extra `while(this->reduce()){}` did help catch extra bits. this means valid reductions aren't happening because we're running out of tokens. if you artificially add some dead tokens, you'll probably dance around the problem. i thought perhaps reduction orders were wrong, but that didn't seem to help.

i didn't investigate too far, but it id look like a:
token{, if-statement-non-expression,token} was not being reduced to a block. not really sure why not. perhaps if an if-statement does move to the final tree it should return true even though it didnt really get reduced?

**Comment by harrand** [comment]
Pretty sure I've now fixed this at e5a89a88ffd4ea70206ffee478843249aa0702f9
